src/traits.rs

```diff
@@ -99,27 +109,42 @@ impl<M, O, N> Step<M, O, N> {
         self
     }
 
-    /// Converts `self` into a step of another type, given conversion methods for output and
-    /// messages.
-    pub fn map<M2, O2, FO, FM>(self, f_out: FO, f_msg: FM) -> Step<M2, O2, N>
+    /// Converts `self` into a step of another type, given conversion methods for output, faults,
+    /// and messages.
+    pub fn map<M2, O2, F2, FO, FF, FM>(
+        self,
+        f_out: FO,
+        f_fail2fail: FF,
+        f_msg: FM,
+    ) -> Step<M2, O2, N, F2>
     where
+        F2: Fail,
         FO: Fn(O) -> O2,
+        FF: Fn(Fault<N, F>) -> Fault<N, F2>,
```

Comment:
We should only have to specify an `Fn(F) -> F2` here: The node ID must never change in this conversion. (Also in the other methods.)

Reply:
Sorry, I don't quite know how to do this one. So the general idea is that `FF` should only work on `Fail` and return `Fail` without ability to touch node id, correct?

If I change the line here (and below it in `extend_with`) to `FF: Fn(F) -> F2`, it seems to break things in some other places (because of type mismatch). Is correcting them one by one the way to go or should I do this in a different way?

Reply:
Yes, I'm afraid they have to be fixed one by one, sorry. :grimacing: 
I imagine something like this:
* Add something like `fn map(self, FF) -> Fault<N, F2> where FF: Fn(F) -> F2` to `Fault<N, F>`, and similarly to `FaultLog`, so that below in line 128 you can just write `fault_log: self.fault_log.map(f_fail2fail)`—and actually I'd rename the latter to `f_fault`.
`Fault::map` would just return `Fault { node_id: self.node_id, kind: f_fault(self.kind) }`.
* Wherever the code currently says something like
```rust
step.extend_with(other_step, |flt| Fault::new(flt.node_id, f(flt.kind)), to_msg);
```
it should be changed to just:
```rust
step.extend_with(other_step, f, to_msg);
```